DRILL-5735: UI options grouping and filtering & Metrics hints

kkhatua opened the pull request:
(Note: DRILL-4699 is also resolved in this)
Additional details, like the description is provided as well in a JScript lookup map. This helps reduce the need for the server to constantly recreate the entire page with the description details, as the client browser can fill in these details.
Developers will be expected to update the description as old/new options are introduced or deprecated.

kkhatua commented:
Sample snapshot when filtering with one of the quick filters being clicked:
![image](https://user-images.githubusercontent.com/4335237/40321599-6c7b860c-5ce4-11e8-8131-fa58f96f0c2c.png)

kkhatua commented:
@dvjyothsna  / @sohami / @parthchandra can one of you please review this?

dvjyothsna commented:
This is really handy. One more thing If I'm not wrong, I think John Omernick was asking for description for options but I think he wanted it as a another table or something. But I think this will help him.

kkhatua commented:
Oh.. good! I am running out of code review volunteers. 
@johnomernik can you review this PR?

kkhatua commented:
@JohnOmernik I agree. Single-source of truth for descriptions will help a long way. 
A similar ask for `sys.functions` is something that @arina-ielchiieva  had pointed to when we introduced syntax highlighting. That, of course, is more trickier due to dynamic UDFs. 
We can try to address that separately as a combined commit. For now, does this look sufficient within the context of the JIRA ?

JohnOmernik commented:
Yes, I like what I am seeing and in the context of the JIRA it looks good.

JohnOmernik commented:
+1

dvjyothsna commented:
Very good enhancement Kunal. LGTM +1

JohnOmernik commented:
So on the idea of UDFs, I don't think UDFs that are NOT shipped with the
product should be a concern of ours.  Any UDF that ships with Drill needs
to be in the sys.functions.

Perhaps it would be cool to have a way to register UDFs in sys.function so
that folks who have UDFs that want to use  sys.functions as a documentation
store can.  Basically, a way in the UDF to add "company only" UDFs to the
sys.functions?

John

On Mon, May 21, 2018 at 2:04 PM, Kunal Khatua <notifications@github.com>
wrote:

> @JohnOmernik <https://github.com/JohnOmernik> I agree. Single-source of
> truth for descriptions will help a long way.
> A similar ask for sys.functions is something that @arina-ielchiieva
> <https://github.com/arina-ielchiieva> had pointed to when we introduced
> syntax highlighting. That, of course, is more trickier due to dynamic UDFs.
> We can try to address that separately as a combined commit. For now, does
> this look sufficient within the context of the JIRA ?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/apache/drill/pull/1279#issuecomment-390751044>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AB1zUa8pjkYSCQ42g6pKwkZEEk9JyxRCks5t0w-6gaJpZM4UGO84>
> .
>

arina-ielchiieva commented:
@kkhatua though the idea of the improvement is nice but having options description in js file is not a good, looks like a hack. :) We should find a way to add description in drill-module.conf file or other approach you can think of. Thus description will be loaded along with the options defaults at start up and thus it would be easily added to the options page, as well as in sys.options table as part of description column,

arina-ielchiieva commented:
@kkhatua, we don't have that many options to have significant load on web server. I understand you want to make options page user-friendly but maintaining options descriptions separately from options is not a good idea. Developers definitely won't remember to update js file when adding new options. So at some point we'll end up with stale descriptions. I am sorry but I would rather have less-user friendly options page then implemented this way. Until we don't have better solution, you can add link to the Drill documentation where options are described on options page.

JohnOmernik commented:
What if the options were loaded into the JS form on first connection to the
web UI and cached? These settings don't change much, so setting a longer
cache timeout and refreshing cache on change would not be a resource drain.
This could be a file (maybe even JS like it is now) but not part of the
code base, it's generated from the system of record.

On Tue, Jun 5, 2018, 2:29 AM Arina Ielchiieva <notifications@github.com>
wrote:

> @kkhatua <https://github.com/kkhatua>, we don't have that many options to
> have significant load on web server. I understand you want to make options
> page user-friendly but maintaining options descriptions separately from
> options is not a good idea. Developers definitely won't remember to update
> js file when adding new options. So at some point we'll end up with stale
> descriptions. I am sorry but I would rather have less-user friendly options
> page then implemented this way. Until we don't have better solution, you
> can add link to the Drill documentation where options are described on
> options page.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/apache/drill/pull/1279#issuecomment-394609967>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AB1zUTWKtmKsAcEbMmTwqmBwRlTIIjKaks5t5jNlgaJpZM4UGO84>
> .
>

parthchandra commented:
There is a PR for DRILL-4699 (adding option descriptions) which, sadly, is a bit out of date. An advantage of using that PR as the base for the UI would be that the same descriptions would also be available in sqlline.

kkhatua commented:
I would like to have the work for descriptions be **delinked** from this issue, which is a usability experience. 
The issue pointed out is related to maintaining of these (and other) descriptions in a coherent state across the UX, the backend code and and documentation. We already have JIRAs - DRILL-4699 & DRILL-3988 .

parthchandra commented:
IMHO, code maintainability and coherence are perfectly valid requirements from a reviewer. In particular, having help text being out of date or inconsistent is a fatal problem for user facing code. If you're inclined, we can work on updating DRILL-4699 together.

ilooner commented:
Since it looks like reviewers have requested additional work to be done on this change, I have removed the ready-to-commit label from the Jira.

arina-ielchiieva commented:
Looks like Kunal wants to separate out sorting & search from descriptions work. I think this is acceptable.

kkhatua commented:
Looks like only @arina-ielchiieva is in agreement with me on having this done incrementally :)

@parthchandra , if updating DRILL-4699 (#536) is not too complex and time-consuming, we can follow that route. I can create a dynamic one-time JScript to be generated during the web server startup, and serve the file from there.

parthchandra commented:
Sounds good to me.

kkhatua commented:
@parthchandra 
Rebased and updated with a resource file as the primary source of truth. The Webserver will now create a JavaScript on the fly and serve that to browsers which can use it to render the information.

parthchandra commented on exec/java-exec/src/main/java/org/apache/drill/exec/server/rest/WebServer.java:
```diff
@@ -408,5 +448,47 @@ public void close() throws Exception {
+  private void generateOptionsDescriptionJSFile() throws IOException {
+    //Obtain properties from Resource file
+    Properties optionsDescriptions = new Properties();
+    InputStream resrcInputStream = this.getClass().getClassLoader().getResourceAsStream(OPTIONS_DESCRIBE_PROPERTIES);
```
This is nice, but using a ResourceBundle would be even better since that will allow the descriptions to be localized without code change.

kkhatua replied:
Let's skip localization for now :)
I'll work towards setting up the PR for DRILL-4699's PR as well, which would be a minor patch over this.

arina-ielchiieva commented on exec/java-exec/src/main/java/org/apache/drill/exec/server/rest/WebServer.java:
```diff
@@ -96,6 +107,23 @@
+        generateOptionsDescriptionJSFile();
+      } catch (IOException e) {
+        // TODO Auto-generated catch block
+        e.printStackTrace();
```
Please remove this and throw appropriate exception.

kkhatua replied:
Ok.

arina-ielchiieva commented on exec/java-exec/src/main/java/org/apache/drill/exec/server/rest/WebServer.java:
```diff
@@ -408,5 +448,47 @@ public void close() throws Exception {
+    File optionsDescriptionFile = new File(getTmpJavaScriptDir(), OPTIONS_DESCRIBE_JS);
+    optionsDescriptionFile.deleteOnExit();
+    logger.info("Will write {} descriptions to {}", numLeftToWrite, optionsDescriptionFile.getAbsolutePath());
+    BufferedWriter writer = new BufferedWriter(new FileWriter(optionsDescriptionFile));
```
Is try with resources possible here?

kkhatua replied:
SHould be. I'll incorporate that change.

arina-ielchiieva commented on exec/java-exec/src/main/resources/options.describe.properties:
```diff
@@ -0,0 +1,106 @@
+store.parquet.block-size=Sets the size of a Parquet row group to the number of bytes less than or equal to the block size of MFS, HDFS, or the file system.
+store.parquet.compression=Compression type for storing Parquet output. Allowed values: snappy, gzip, none
+store.parquet.enable_dictionary_encoding=For internal use. Do not change.
+store.parquet.dictionary.page-size=
```
description?

kkhatua replied:
Picked this up from the website, so I guess this slipped through. :)

arina-ielchiieva commented on exec/java-exec/src/main/java/org/apache/drill/exec/server/rest/WebServer.java:
```diff
@@ -408,5 +448,47 @@ public void close() throws Exception {
       embeddedJetty.stop();
     }
+    //Deleting temp directory
+    FileUtils.deleteDirectory(getTmpJavaScriptDir());
```
Will this fail if we unable to delete temp dir?

kkhatua replied:
My guess is that deleting a non-empty directory without this would not work. This should recursively go and clean up.

arina-ielchiieva replied:
I meant what if we fail to delete temp directory? May be use delete quietly.

arina-ielchiieva replied:
Again defending `getTmpJavaScriptDir()` usage is incorrect. Consider the case when server start will fail before generating js temp dir, so in close method you'll call `getTmpJavaScriptDir()` and it will generate js options since it did not do it at start up since all we need is to delete the directory.

kkhatua replied:
I wanted to generate the JS lazily, hence this approach. I'll make the necessary change and generate it on startup.

arina-ielchiieva commented on exec/java-exec/src/main/java/org/apache/drill/exec/server/rest/WebServer.java:
```diff
@@ -96,6 +107,23 @@
+  private static File tmpJavaScriptDir;
+
+  public File getTmpJavaScriptDir() {
+    if (tmpJavaScriptDir == null) {
```
How many times this would be called? I am just wondering why we have this null check.

kkhatua replied:
I wanted to bundle the creation as a part of the dir creation, so this check is for that.

arina-ielchiieva commented on exec/java-exec/src/main/java/org/apache/drill/exec/server/rest/WebServer.java:
```diff
@@ -96,6 +107,23 @@
 
   private Server embeddedJetty;
 
+  private static File tmpJavaScriptDir;
```
Can we store it as local variable?

kkhatua replied:
Being a single instance, I thought a static would be preferable.

kkhatua commented:
@arina-ielchiieva I should be able to make most of the changes you've asked for.
We'll need to inform devs about this as well. 
However, i have a second problem related to DRILL-4699, which can be easily built off this. The descriptions are very lengthy and break the console formatting when applying to the `sys.options` table. To make things more confusing, there is now a `sys.options_val` table as well ! I'm in favor of replacing the `sys.options` with the `sys.options_val` implementation. So, I'm going to slightly revise this PR in anticipation of an additional PR for DRILL-4699, where the same/similar resource can be used for displaying a shorter description. (limiting to <110 chars doesn't seem to break formatting)

kkhatua commented:
@arina-ielchiieva  / @parthchandra 

Modified the PR with additional work:
1. The latest commit now carries an implementation for description in `sys.options_val` table (DRILL-4699), making PR #536 obsolete
2. The descriptions-carrying resource file is loaded by the `ExtendedOptionIterator` and made accessible at a class level to the `WebServer` instance
4. The full description is to be rendered in the WebUI (i.e. http://drillbit:8047/options )
5. The short description (if available) is shown in the System table, and truncated with suffixing of ellipsis to indicate more content. (See bottom). This is required so as to ensure that the formatting within SqlLine is maintained.
```
0: jdbc:drill:drillbit=kk127.qa.lab> select * from sys.options_val where name like 'planner.p%' or name like 'planner.width.%' or name like '%cpu%' order by name;
+--------------------------------------------+----------+-------------------+-------+----------+--------------+-----------------------------------------------------------------------------------------------------------------+
|                    name                    |   kind   | accessibleScopes  |  val  |  status  | optionScope  |                                                   description                                                   |
+--------------------------------------------+----------+-------------------+-------+----------+--------------+-----------------------------------------------------------------------------------------------------------------+
| planner.cpu_load_average                   | FLOAT    | ALL               | 0.7   | DEFAULT  | BOOT         |                                                                                                                 |
| planner.parser.quoting_identifiers         | VARCHAR  | ALL               | `     | DEFAULT  | BOOT         |                                                                                                                 |
| planner.partitioner_sender_max_threads     | BIGINT   | ALL               | 8     | DEFAULT  | BOOT         | Upper limit of threads for outbound queuing.                                                                    |
| planner.partitioner_sender_set_threads     | BIGINT   | ALL               | -1    | DEFAULT  | BOOT         | Overwrites the number of threads used to send out batches of records. Set to -1 to disable. Typically not c...  |
| planner.partitioner_sender_threads_factor  | BIGINT   | ALL               | 2     | DEFAULT  | BOOT         | A heuristic param to use to influence final number of threads. The higher the value the fewer the number of...  |
| planner.producer_consumer_queue_size       | BIGINT   | ALL               | 10    | DEFAULT  | BOOT         | How much data to prefetch from disk in record batches out-of-band of query execution. The larger the queue ...  |
| planner.width.max_per_node                 | BIGINT   | ALL               | 0     | DEFAULT  | BOOT         | Max number of threads that can run in parallel for a query on a node.                                           |
| planner.width.max_per_query                | BIGINT   | ALL               | 1000  | DEFAULT  | BOOT         | Same as max per node but applies to the query as executed by the entire cluster. For example, this value mi...  |
+--------------------------------------------+----------+-------------------+-------+----------+--------------+-----------------------------------------------------------------------------------------------------------------+
8 rows selected (0.797 seconds)
```
6. Review comments have been addressed:
  a. JavaScript is written using try-resources
  b. tmpJavaScriptDir is a class instance member
  c. Missing description has been added. For testing purposes, a `planner.width.max_per_node.short` entry also exists (as shown above)

Going forward, **ALL** developers need to introduce the description of their introduced options.

arina-ielchiieva commented on exec/java-exec/src/main/java/org/apache/drill/exec/store/sys/ExtendedOptionIterator.java:
```diff
@@ -52,14 +55,44 @@
 
   private final OptionManager fragmentOptions;
   private final Iterator<OptionValue> mergedOptions;
+  private HashMap<OptionValue.Kind,String> typeMapping;
```
Can we use more generic `Map`, plus space after comma.

kkhatua replied:
Ok, just moved the method out from where it originally was. Will make the change

arina-ielchiieva commented on exec/java-exec/src/main/java/org/apache/drill/exec/store/sys/ExtendedOptionIterator.java:
```diff
@@ -52,14 +55,44 @@
+  //Load & Provide a descriptions map
+  public static Properties getDescriptionMap() {
+      if (descriptionMap.isEmpty()) { //Expect a one time load
+        synchronized (descriptionMap) {
```
Re-check for empty after sync (DCL).

arina-ielchiieva commented on exec/java-exec/src/main/java/org/apache/drill/exec/store/sys/ExtendedOptionIterator.java:
```diff
@@ -52,14 +55,44 @@
+          try (InputStream optionDescInputStream = OptionIterator.class.getClassLoader().getResourceAsStream(OPTIONS_DESCRIBE_PROPERTIES)) {
+            descriptionMap.load(optionDescInputStream);
+          } catch (IOException e) {
+            logger.warn("Unable to load descriptions from {}", OPTIONS_DESCRIBE_PROPERTIES);
```
Add exception to logger.

kkhatua replied:
My bad. I think I missed adding it in the refactor.

arina-ielchiieva commented on exec/java-exec/src/main/java/org/apache/drill/exec/store/sys/ExtendedOptionIterator.java:
```diff
@@ -52,14 +55,44 @@
   public ExtendedOptionIterator(FragmentContext context, boolean internal) {
     fragmentOptions = context.getOptions();
-    final Iterator<OptionValue> optionList;
+    preference = new HashMap<OptionScope, Integer>() {{
```
You are actually extending the HashMap, not sure this is preferable. Please use standard way of adding values.

kkhatua replied:
This is just refactored code that I moved from elsewhere.  Since the entire statement is creating a constant map, this encapsulates that. Should I still make the change?

arina-ielchiieva replied:
I would prefer if you would :)

kkhatua replied:
Done

arina-ielchiieva commented on exec/java-exec/src/main/java/org/apache/drill/exec/store/sys/ExtendedOptionIterator.java:
```diff
@@ -115,14 +143,20 @@ public boolean hasNext() {
+      status = Status.BOOT;
+    } else {
+      final OptionValue def = fragmentOptions.getDefault(value.name);
+      if (value.equalsIgnoreType(def)) {
```
Use ternary operator.

kkhatua replied:
👍 Ok. Was avoiding refactoring the original code, just I copied as is from OptionsIterator. :)

arina-ielchiieva commented on exec/java-exec/src/main/java/org/apache/drill/exec/store/sys/ExtendedOptionIterator.java:
```diff
@@ -134,19 +168,36 @@ public ExtendedOptionValueWrapper next() {
+      this.description = getShortDescription(name);
+    }
+
+    private String getShortDescription(String name) {
```
Please add comment why we might need short description.

kkhatua replied:
👍

arina-ielchiieva replied:
Has comment been added?

kkhatua replied:
Looks like I missed this. Done in the latest commit.

arina-ielchiieva commented on exec/java-exec/src/main/java/org/apache/drill/exec/store/sys/SystemTable.java:
```diff
@@ -31,6 +31,7 @@
  * </p>
  */
 public enum SystemTable {
+  @Deprecated
```
Why we need to deprecate?

kkhatua replied:
I want to actually replace the sys.options completely with sys.options_val , but I broke tests. Hoping to discourage the usage this way.

parthchandra replied:
It should really be the other way around; i.e deprecate sys.options_val and retain sys.options. Options_val coalesces the xxx_value columns to show only one value column which is desirable. But it also shows only the value in effect for the user which IMO is less useful. 
@amansinha100 had a good suggestion which was to have one large sys.options table but provide appropriate views for different use cases.

kkhatua replied:
I don't think the challenge is so much in deprecating either. I actually have a commit that renames the two tables as `sys.options_old` and s`ys.options`. The concern seems to be around breaking backward compatibility tests.
Are you suggesting we show default values in addition to changed value??

parthchandra replied:
I see inconsistent behaviour for sys.options. If I override an option at the session level, two rows show up in the table, one for the default and one for the overridden (sys.options_val always shows only one). If, however, I override an option at the system level only one row shows up in the table, the overridden one. 
Ideal would be to show both, but also coalesce the columns for the value (as in options_val).
I think we should log a separate Jira for that.

kkhatua replied:
Interesting. I've never seen that whenever I've modified at a session level. Could you file a JIRA for that?

kkhatua replied:
The bug is not from this commit, so will ignore this for now. 
Should I leave this deprecated or remove it?

arina-ielchiieva replied:
What is the decision on this part?

kkhatua replied:
I've removed the deprecated flag for now. I logged [DRILL-6684](https://issues.apache.org/jira/browse/DRILL-6684)  for this purpose.

arina-ielchiieva commented on exec/java-exec/src/main/java/org/apache/drill/exec/server/rest/WebServer.java:
```diff
@@ -194,6 +220,16 @@ private ServletContextHandler createServletContextHandler(final boolean authEnab
+    //Add Local path resource (This will allow access to dynamically created files like JavaScript)
+    final ServletHolder dynamicHolder = new ServletHolder("dynamic", DefaultServlet.class);
+    //Skip if unable to get a temp directory (e.g. during Unit tests)
+    if (getTmpJavaScriptDir() != null) {
```
Server start won't be called several times. I guess just calling getTmpJavaScriptDir is sufficient and hen refer to tmpJavaScriptDir. And do null check during shutdown.

kkhatua replied:
Ok.

kkhatua replied:
@arina-ielchiieva  I'm unable to reply to your comment on line 452, so doing it here. 
```
arina-ielchiieva on Jun 15  Contributor
I meant what if we fail to delete temp directory? May be use delete quietly.
```
The temp directory is already marked as `deleteOnExit()` , so the JVM should also do that during shutdown. However, it didn't seem to work always, so an explicit delete is fine.

```
arina-ielchiieva on Jun 15  Contributor
Again defending getTmpJavaScriptDir() usage is incorrect. Consider the case when server start will fail before generating js temp dir, so in close method you'll call getTmpJavaScriptDir() and it will generate js options since it did not do it at start up since all we need is to delete the directory.
```
The creation of the JS options is a negligible cost, as long as we are cleaning up correctly.

arina-ielchiieva commented on exec/java-exec/src/main/java/org/apache/drill/exec/server/rest/WebServer.java:
```diff
@@ -408,5 +444,46 @@ public void close() throws Exception {
+  }
+
+  /* Static wrapper text for JavaScript content */
+  private static final String FILE_CONTENT_HEADER =
```
Consider having some template where you'll insert only options and descriptions rather then concatenating js file from string pieces.

kkhatua replied:
You mean a freemarker template? That would involve the webserver regenerating the file on every request

kkhatua replied:
I could use a standalone template file (say, `options.describe.js.template`) and then construct the the final file with a copy from this and inject the options and descriptions. I think that makes it cleaner and helps keep JavaScript code outside of the Java code.

parthchandra commented:
Why not add the description in sys.options? It is the more commonly used table, and also the logical place to look for options and their descriptions. (And yes, I know that it provides less space for the description; but you just need to wrap around and live with the slightly ugly look). 
I think we also need to merge the sys.options and sys.options_val tables (there is no need to have both), but that is a topic for a different Jira/PR.

kkhatua commented:
@parthchandra 
I actually have my original PR with that. But I seemed to be breaking some backward compatibility tests. Not sure why a sys.options table needs backward compatibility, with the exception of options persisted in ZK, but that is the case.  That's why ( @arina-ielchiieva  ), for now, I marked the `sys.options` as deprecated.

kkhatua commented:
@arina-ielchiieva I'm going to split this PR into 2 parts, with the UI part going in first. The changes to the `sys.options` table will follow, after I figure out a fix to Parth's reported bug.  Would that be sufficient for a +1 ?

Agirish commented:
This is really cool! Love the UI. Not sure what's with the new options_val table - is this going to replace sys.options? Didn't get the reasoning behind that.

kkhatua commented:
Ok. Just checked. The bug reported by Parth has been around even before my commit. As such, I don't see the value in splitting the commit then. 
@arina-ielchiieva , I'll try to address the comments in your review.

@Agirish the `options_val` table was introduced to allow a unified value column instead of different column values for for each possible type (e.g. bool_val , string_val , ...) most of which would be null.

arina-ielchiieva commented on exec/java-exec/src/main/java/org/apache/drill/exec/store/sys/ExtendedOptionIterator.java:
```diff
@@ -116,14 +146,16 @@ public boolean hasNext() {
+      status = Status.BOOT;
+    } else {
+      final OptionValue def = fragmentOptions.getDefault(value.name);
+      status = (value.equalsIgnoreType(def) ? Status.DEFAULT : Status.CHANGED );
```
Please remove space before `)`.

kkhatua replied:
Done

arina-ielchiieva commented on exec/java-exec/src/main/resources/options.describe.properties:
```diff
@@ -0,0 +1,107 @@
+#
```
Since we have now options.describe.properties file, how did we agree to enforce developers to add descriptions?
Will project build fail if description was not added? Or maybe instead of using properties file, we can make `OptionValidator` to accept description along with option name, thus all options will be required to have description? And we won't have to duplicate options names in options.describe.properties?

kkhatua replied:
For future additions, we could do this, but I didn't want to right away change the existing OptionValidator usages. I'll need to work with @bbevens to fill up the properties file for now, since total list of options are 150+.

parthchandra replied:
We can log a new JIRA for updating the Options validator so that all options have a description and move on.

arina-ielchiieva replied:
Sorry guys but I think it's better not to introduce `options.describe.properties` file at all. Updating Options validator is not that hard, for those options we don't have a description prepared, just leave blanks.

kkhatua replied:
@arina-ielchiieva the purpose of the PR was a UX enhancement. The backend is a temporary change until the proposed change to the OptionValidator is implemented later. Time to change existing OptionValidators would require significant time. Introducing the complete change now, by leaving blanks defeats the purpose of the UX enhancement.

arina-ielchiieva replied:
@kkhatua usually there is nothing more persistent then temporary :)
Current list of options in properties files does not contain full list of system / session options. It means that the rest would have empty description, right? So I am not sure how this is better then updating OptionsValidators. Expanding the constructor and adding descriptions to options, does not seem to take a significant time, might be not the most interesting part of the development but still.

kkhatua replied:
Yes, some options have empty descriptions because I didn't find them on the website. I'm looking to get in the UX enhancement originally intended for 1.14.0. I also wanted to (separately) discuss whether changes to the contructor is suitable or introduction of annotations.

arina-ielchiieva replied:
I would go with constructor.

kkhatua commented:
@arina-ielchiieva please review.

arina-ielchiieva commented:
@kkhatua two questions:
1. Will `sys.options` return descriptions along previous information (this can be useful)?
2. Do we need `options.describe.properties` ? We can get all descriptions from option manager now.

kkhatua commented:
@arina-ielchiieva 
1. The descriptions are, as of now, part of `sys.options_val` (and sys.internal_options_val`) due to the width of `sys.options` . I intend to use [DRILL-6684](https://issues.apache.org/jira/browse/DRILL-6684) to swap the names of the two system tables. That is a bigger exercise because unit tests, etc need to be adjusted accordingly.
2. I've dropped it from the last commit. 

I'll work with Bridget and other developers to update their features' property descriptions. As of now, we've barely got half of these covered.

arina-ielchiieva commented:
@kkhatua please squash the commits.

kkhatua commented:
Done. Rebased and squashed. Cleared unit and functional tests.

arina-ielchiieva commented:
@kkhatua please resolve the conflicts.

kkhatua commented:
@arina-ielchiieva  Done.

kkhatua commented:
Rebased again due to Guava shading.

sohami commented:
@kkhatua - please make PR and JIRA title same.
